records/2024-06-04.manifests-kustomize-cicd.md

```diff
@@ -0,0 +1,183 @@
+
+#### Different Release Process For Staging and Production
+
+Manifests code is not actually changed when doing a release to staging, thus making it difficult to track changes and perform effective rollbacks.
+
+#### Different Release Process Between Terraform and Manifests
+
+Terraform does not use the notify-pr-bot to generate automatic PRs for Terraform changes. This creates unnecessary toil for the ops lead when doing Terraform releases.
+
+#### Staging Releases Get Trampled
+
+Because we do not actually update any manifests code when doing an individual application rollout, we lose all context of which version we should be using when we do a global manifest rollout with merge to main (staging). To overcome this, the staging Kustomize code is set to always use the "latest" tag. This means when we apply Kustomize the explicitly versioned docker tag is overwritten with "latest". This works because the docker build and push job for each application tags as both latest and the github sha. It is however not ideal since it becomes difficult to track versions in staging, and also creates a delta between how staging and production releases are handled through release pipelines.
+
+This issue is a problem regardless of whether we use Kustomize or Helmfile.
+
+#### No Ability to Preview Manifests Changes
+
+Because Kustomize lacks any sort of "diff" functionality, we are unable to know exactly what will happen when we release to staging or production. This creates uncertainty on releases and introduces the risk of deploying unwanted changes. There have been several incidents related to this issue.
+
+#### Self-Hosted Github ARC Runners
+
+The self-hosted Github ARC runners are currently required since the EKS cluster is private. This introduces additional complexity, and a new failure point. There have already been several incidents where the ARC runner failed to run, delaying our releases.
+
+## Proposed Solution
+
+### Step 1: Remove Github ARC Runners In Favor of VPN
+
+We can create a secondary VPN endpoint in AWS that uses certificate based authentication instead of SAML based. This will allow standard Github runners to remotely connect to our Kubernetes API, and remove the necessity of internal Github ARC Runners.
+
+There would be an increased cost to this solution, but it could be mitigated by reducing the number of connected subnets to each VPN from 3 to 1. Tests have been run to verify that connectivity to the other subnets remains.
+
+#### Problems Solved
+
+- Using self hosted Github runners introduces complexity and additional work.
+
+#### Pros
+
+- Less complexity involved for managing releases
+- Completely removes the requirement of Github ARC
+- Removes potential security risks inherent to Docker in Docker
+- Removes the "Chicken and Egg" scenario where we can't use Github Actions to deploy Kubernetes code until the Kubernetes cluster is set up.
+
+#### Cons
+
+- Adds an additional VPN endpoint to manage
```

Comment:
Are there cost considerations too?

Reply:
We lowered the number of subnet connections on both the main VPN and GHA VPN so that there will not be a cost impact.